### PR TITLE
[Draft] Design Doc: Data Pruning

### LINKED ISSUE #74: [Design Doc] Storage
Design doc around pruning historical state

Basics
- While people could run archive nodes, we want to make it such that there is limited reason to do so.
- Why? Cost of running nodes. For comparison, Ethereum Archive node is at >10TB while a full node is < 1TB.
- Not forcing, but not required to read state (and any historical state could still be validated)
- Outcomes:
  - We store more information in state than otherwise need to, but total storage is lower.
  - Lower gossip data requests. Effectively full nodes have all the data you could need.
  - Schema Messages don't have to be permanent
- Trust model changes
  - Ethereum you can trust (like most people) the consensus, but you could also do your own validation and run an archive node.
  - MRC still offers that choice, but only from (about) the point in time that you started archiving. If you didn't start early, then you might have to trust that the consensus had the correct state at the point you start archiving.

### PR DESCRIPTION
**Pull request by wilwade**
Design doc around pruning and data storage
Closes #74

### CODE REVIEW

**Review by aramikm — APPROVED** [review]
I think this sums up the problem nicely

**Review by saraswatpuneet — APPROVED** [review]

File: designdocs/pruning_data.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+If you want to pull all 10TB+ of data down (after several restarts), you can test to make sure that each state transition of each transaction and block were applied correctly and that it matches the rest of the world.
+Or you could just be a "full node" pull the block headers, the current state, and the past few hundred blocks for about 1TB of data.
+The full node version trusts that the information prior to the few hundred blocks is good.
+
+MRC is structured such that we can survive without any historical block data beyond the last few hundred.
+Effectively, we don't require archive nodes.
+Someone could run one, or at least run one that starts from the time they started it, but the network functions without it.
+Also as we still maintain block headers, a historical block is still able to be validated.
+
+Effectively, after a given point in time, we trust that a given block was validated correctly by consensus and by the [relay chain](https://polkadot.network/technology/).
```
Comment (shannonwells) [inline designdocs/pruning_data.md:32]:
```suggestion
This means that after a given point in time, we trust that a given block was validated correctly by consensus and by the [relay chain](https://polkadot.network/technology/).
```

File: designdocs/pruning_data.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+The relay chain's secure validation provides a greater degree of security for MRC without having to rely on the MRC's collator nodes.
+
+The outcome of this choice does mean that slightly more information is held in active state.
+Messages store additional information that would normally be stored in the transaction alone, because that transaction might not be retrievable after some time.
+The total storage require is still reduced.
```
Comment (shannonwells) [inline designdocs/pruning_data.md:37]:
```suggestion
The total storage required is still reduced.
```

File: designdocs/pruning_data.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+## Active State Pruning
+
+Since we have historical block pruning, we have to be extra careful what we prune from active state.
+Any information we prune, must be expected to be lost forever.
```
Comment (shannonwells) [inline designdocs/pruning_data.md:42]:
```suggestion
Any information we prune, must be expected to be lost forever, however,
```

File: designdocs/pruning_data.md [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+However that is also a benefit.
```
Comment (shannonwells) [inline designdocs/pruning_data.md:43]:
```suggestion
that is also a benefit.
```

File: designdocs/pruning_data.md [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+MRC can be structured such that selected information is forgotten.
+
+We leverage this with allowing schemas to be configured with a time to live or [TTL](https://en.wikipedia.org/wiki/Time_to_live) for messages.
```
Comment (shannonwells) [inline designdocs/pruning_data.md:46]: Unclear what "this" means.

**Review by shannonwells — CHANGES_REQUESTED** [review]
This isn't really a design document; it's too general and high level and is more of a starting point for discussions.  You could not generate implementation stories from this.  I'm not sure where it belongs instead, it just should not be in designdocs.

**Pull request closed without merging**